#67 소셜 로그인 기능 개발

taesikk opened the pull request:
## Sourcery 요약

Google OAuth 및 QR 코드 기반 인증을 포함한 소셜 로그인 기능을 구현하고, 백엔드에 Firebase 및 Redis를 통합하며, 토큰 유틸리티를 사용하여 API 호출 보안을 강화하고, 일관된 테마 적용 및 동적 리소스 경로를 위해 스타일링 및 빌드 스크립트를 개선했습니다.

### 새로운 기능:
- 프론트엔드 및 백엔드에서 Firebase를 활용한 Google OAuth 소셜 로그인 플로우 구현
- 세션 생성, 폴링(polling) 및 모바일 스캔 페이지를 포함한 QR 코드 로그인 플로우 추가
- 인증된 AJAX 요청 처리를 위한 `tokenAjax` 및 `tokenAjaxToJson` 헬퍼 도입

### 개선 사항:
- 레이아웃 및 로그인 페이지 전반에 걸쳐 일관된 다크 테마 UI를 위한 믹스인(mixins) 및 업데이트된 변수를 사용하여 SCSS 구조 개편
- Firebase Admin 및 Redis 스토리지를 통합하여 `LoginController` 및 `LoginService`에 소셜 로그인, 로그아웃 및 QR 세션 엔드포인트 확장
- Firebase 토큰을 사용하여 `ChatRoomController`에서 인증을 강화하고, 참여 응답에 사용자 정보 포함
- 동적 `convert_path.json` 매핑을 사용하여 Electron 데스크톱 경로 확인 기능 강화

### 빌드:
- 동적 상대 정적 경로를 위한 빌드 스크립트 경로 변환기 개선
- Node.js 프론트엔드 및 Spring Boot 백엔드에 Firebase 종속성 추가

### 테스트:
- 새로운 소셜 사용자 키 접두사를 포함하도록 `RedisServiceImplTest` 업데이트

### 기타 작업:
- 로그인 및 QR 페이지의 프론트엔드 템플릿과 SCSS 모듈을 전용 디렉토리 아래로 재구성

<details>
<summary>Original summary in English</summary>

## Summary by Sourcery

Implement social login features including Google OAuth and QR code-based authentication, integrate Firebase and Redis in backend, secure API calls with token utilities, and enhance styling and build scripts for consistent theming and dynamic resource paths

New Features:
- Implement Google OAuth social login flow with Firebase in frontend and backend
- Add QR code login flow including session generation, polling and mobile scanning pages
- Introduce tokenAjax and tokenAjaxToJson helpers for authenticated AJAX requests

Enhancements:
- Revamp SCSS structure with mixins and updated variables for consistent dark-themed UI across layouts and login pages
- Extend LoginController and LoginService with social login, logout and QR session endpoints integrating Firebase Admin and Redis storage
- Enforce authentication in ChatRoomController using Firebase tokens and include user info in join responses
- Enhance Electron desktop path resolution using dynamic convert_path.json mapping

Build:
- Improve build-scripts path-converter for dynamic relative static paths
- Add Firebase dependencies to Node.js frontend and Spring Boot backend

Tests:
- Update RedisServiceImplTest to cover new social user key prefixes

Chores:
- Reorganize frontend templates and SCSS modules for login and QR pages under dedicated directories

</details>

Commit:
#67 firebase 프론트 의존성 추가

Commit:
#67 프론트 밑작업

Commit:
#67 firebase 이용한 프론트 토큰 발급 로직 추가

Commit:
#67 컨트롤러 수정 및 백엔드 작업중

Commit:
#67 localStorage 세팅 작업

Commit:
#67 레디스 로직 추가 및 토큰 리턴 로직 추가

Commit:
#67 매핑 및 객체 변경

Commit:
#67 firebase 토큰 검증 초기화 로직 추가

Commit:
#67 firebase verify 로직 개발 완료

taesikk committed:
Merge branch 'chatforyou_v2' into feat/67_2

# Conflicts:
#	.gitignore
#	springboot-backend/build.gradle
#	springboot-backend/src/main/java/webChat/controller/LoginController.java
#	springboot-backend/src/main/java/webChat/service/redis/RedisService.java

taesikk committed:
#67 토큰 레디스 저장 및 조회 객체 변경

gahyun-lim777 committed:
#67_2 로그인 페이지 ui 개선

taesikk committed:
#67 로그아웃 기능 추가

gahyun-lim777 committed:
#67_2 로그인 페이지 ui 개선 2

SeJonJ committed:
#67 방 생성 입장 시 로그인 체크 및 입장 후 user name 처리

- css 에서 electron 시 빌드 에러나는 부분 수정
- 방 입장/생성/삭제/조회 등 이벤트 시 버그 수정 및 코드 정리

SeJonJ committed:
#67 QR 로그인 로직 개발(백엔드)

1. QR 로그인 api 개발
2. QR generater 개발
3. QR 세션 redis 저장 로직 개발

SeJonJ committed:
#67_2 QR 로그인 로직 개발(백엔드) :: QR 로그인 코드 정리 및 NickName 중복체크 추가

SeJonJ committed:
#67_2 QR 로그인 로직 개발(프론트) :: QR 로그인 프론트 개발 및 기존 코드 정리

1. qr 로그인을 위한 html 및 js 추가
2. 기존 코드 path 정리

SeJonJ committed:
#67_2 QR 로그인 로직 개발(프론트) :: Electron QR 로그인 빌드 코드 수정

gahyun-lim777 committed:
#67_2 로그인 페이지 :: qr로그인 추가, 대시보드 로그아웃 버튼 uiux 개선

Commit:
#67 로그인, 로그아웃 버튼 기능 추가

SeJonJ committed:
#67 채팅방 닉네임을 로그인 사용자 닉네임으로 설정

1. 게임 및 문자 채팅 시 닉네임도 사용자 닉네임으로 나오도록 개선
2. 게임 음성 인식 개선

taesikk committed:
#67 config 파일 추가

SeJonJ committed:
#67 QR 로그인 처리 개선 및 electron 용 config 값 추가

1. electron 용 QR 로그인 버튼 및 로그아웃 이벤트 추가
2. electron 빌드용 config 값 추가
3. QR 로그인 후처리 개선

SeJonJ committed:
#67 로그인 페이지 리다이렉트 경로에 BASE_URL 추가

SeJonJ committed:
#67 로그아웃 시 url  동작 버그 수정

taesikk committed:
#67 로그인 관련 에러처리 로직 추가

SeJonJ committed:
#67 로그인 후 redirect url 변경 및 logout 시 auth 체크 제거

SeJonJ committed:
Merge branch 'feat/67_2' of https://github.com/SeJonJ/Spring-WebSocket-Chatting into feat/67_2

kidonge commented on nodejs-frontend/static/js/popup/room_popup.js:
```diff
@@ -19,23 +19,23 @@ const RoomPopup = {
-        // 비밀방 모달에서 '입장하기' 버튼 클릭
-        $(document).off('click', '#enterRoomBtn').on('click', '#enterRoomBtn', function(e) {
+        // 일반방 입장
+        $(document).off('click', '.directEnterBtn').on('click', '.directEnterBtn', function(e) {
```
비밀방 입장 이벤트는 요소의 아이디로 이벤트를 등록하였으나 왜 일반방 입장은 classs로 하였는지...
아이디로 변환해주셨으면 좋겠습니다.

SeJonJ replied:
room_popup.js 를 확인하여 class 가 아닌 id 로 이벤트 처리할 수 있도록 수정하였습니다.

kidonge commented on springboot-backend/src/main/java/webChat/controller/ChatRoomController.java:
```diff
@@ -37,13 +41,24 @@ public class ChatRoomController {
+            @RequestBody ChatRoomInVo chatRoomInVo) throws Exception {
+
+        // token 확인
+        FirebaseToken token = new TokenUtils().checkGoogleOAuthToken(authorization);
```
토큰 체크는 공통적으로 많이 사용할 것 같으니 별도의 유틸로 빼두면 좋을 것 같습니다.

SeJonJ replied:
TokenUtils 의 해당 메서드를 static 으로 변경하여 공통화 하였습니다.

kidonge reviewed (commented):
리뷰 확인 완료

SeJonJ committed:
#67 sourcery 코드 리뷰 및 kidonge 코드 리뷰 반영

SeJonJ committed:
Merge branch 'chatforyou_v2' into feat/67_2

# Conflicts:
#	nodejs-frontend/static/js/roomlist/roomlist.js

SeJonJ commented on nodejs-frontend/static/js/login/qr/qrscan.js:
```diff
@@ -0,0 +1,135 @@
+             callbacks: {
+                 signInSuccessWithAuthResult: async (authResult) => {
+                     console.log('Firebase 로그인 성공:', authResult.user.email);
+                     const user = authResult.user;
```
객체 비구조화로 변경완료하였습니다.

SeJonJ commented on nodejs-frontend/static/js/login/qr/qrlogin.js:
```diff
@@ -0,0 +1,230 @@
+            self.sessionId = result.data.sessionId;
+            
+            let qrUrl = result.data.qrUrl;
+            let qrImage = result.data.qrImage;
```
객체 비구조화로 변경완료하였습니다.

SeJonJ commented on nodejs-frontend/static/js/login/qr/qrlogin.js:
```diff
@@ -0,0 +1,230 @@
+            console.log('[QR Login] 세션 생성 성공:', result);
+            self.sessionId = result.data.sessionId;
+            
+            let qrUrl = result.data.qrUrl;
```
객체 비구조화로 변경완료하였습니다.

SeJonJ commented on nodejs-frontend/static/js/login/main.js:
```diff
@@ -83,5 +85,104 @@
+               (window.process && window.process.versions && window.process.versions.electron) ||
+               (window.require && window.require('electron')) ||
+               window.__dirname !== undefined;
+    }
```
객체 비구조화로 변경완료하였습니다.

SeJonJ commented on chatforyou-desktop/build-scripts/lib/path-converter.js:
```diff
@@ -448,11 +470,89 @@ class PathConverter {
     const rules = this.conversionRules[fileType] || [];
     let filePathsConverted = 0;
 
+    // 동적 base href 처리 (convert_path.json 활용)
```
수정불가...너무 많아요ㅠㅠ
난중에는 꼭 확인하겠습니다....

SeJonJ commented on springboot-backend/src/main/java/webChat/service/login/impl/LoginServiceImpl.java:
```diff
@@ -0,0 +1,184 @@
+    public GoogleOAuth checkSocialUser(@NonNull String accessToken, @NonNull String refreshToken, @NonNull String name, @NonNull String email, boolean emailVerified, String photo) {
+        if (!emailVerified) {
+            // TODO 예외처리?? :: googleOAuth.isEmailVerified() 값이 정확히 뭔지 모르겠어요ㅠㅠ
+        }
```
TODO 로 설정해두었으며, 이후 exception 정의 후 한번에 처리할 예정입니다.

SeJonJ committed:
#67 sourcery 코드 리뷰 반영

1. qrlogin 에서 다시 시도 버튼 표시 시 style 대신 jqeury css 변경
2. userdto 에서 builder 에 type 추가 및 type 와 photourl 의 값이 null 인 경우 제외하도록 추가

kidonge reviewed (approved):
수고하셨습니다
